Home Assistant OS leaks private host names to cloudflare DNS service (CVE-2020-36517)

### Describe the issue you are experiencing

Home Assistant OS always gets configured to load-balance between my DNS resolver and Cloudflare's. I already looked under every stone, but there's no way to disable this behavior using configuration options.

So if I use DNS names configured by my router instead of hard-coding IP addresses, HA will switch over to Cloudflare from time to time and leak my internal hostnames to their service. Of course resolution fails in this case, too. Note that this is not caused by the hard-coded fallback option, but by the hard-coded load-balancing option.

One way to avoid this could be blocking 1.1.1.1 and 1.0.0.1 in my firewall, but then HA's DNS resolver goes berserk flooding my firewall logs with multiple connection attempts per second. Also, this will break further whenever developers decide to switch their user base from Cloudflare to a new data collection service.


### What operating system image do you use?

ova (for Virtual Machines)

### What version of Home Assistant Operating System is installed?

7.1

### Did you upgrade the Operating System.

Yes

### Steps to reproduce the issue

1. Insert local hostname in HA config
2. Notice recurring failures in name resolution
3. Notice packets going to 1.0.0.1 and 1.1.1.1


### Anything in the Supervisor logs that might be useful for us?

```txt
-
```


### Anything in the Host logs that might be useful for us?

```txt
Typical output of `ha dns log` when the service gets blocked.
~~~
[INFO] 127.0.0.1:44130 - 10230 "NS IN . udp 17 false 512" NOERROR - 0 30.001023101s
[ERROR] plugin/errors: 2 . NS: dial tcp 1.1.1.1:853: i/o timeout
[INFO] 127.0.0.1:51517 - 36166 "NS IN . udp 17 false 512" NOERROR - 0 30.000661016s
[ERROR] plugin/errors: 2 . NS: dial tcp 1.0.0.1:853: i/o timeout
[INFO] 127.0.0.1:34985 - 13557 "NS IN . udp 17 false 512" NOERROR - 0 30.001216215s
[ERROR] plugin/errors: 2 . NS: dial tcp 1.0.0.1:853: i/o timeout
[INFO] 127.0.0.1:39902 - 40031 "NS IN . udp 17 false 512" NOERROR - 0 30.000821668s
[ERROR] plugin/errors: 2 . NS: dial tcp 1.1.1.1:853: i/o timeout
[INFO] 127.0.0.1:43380 - 37784 "NS IN . udp 17 false 512" NOERROR - 0 30.000649803s
[ERROR] plugin/errors: 2 . NS: dial tcp 1.0.0.1:853: i/o timeout
[INFO] 127.0.0.1:60231 - 1526 "NS IN . udp 17 false 512" NOERROR - 0 30.000941436s
[ERROR] plugin/errors: 2 . NS: dial tcp 1.0.0.1:853: i/o timeout
[INFO] 127.0.0.1:42409 - 17159 "NS IN . udp 17 false 512" NOERROR - 0 30.001298402s
[ERROR] plugin/errors: 2 . NS: dial tcp 1.1.1.1:853: i/o timeout
[INFO] 127.0.0.1:37011 - 7485 "NS IN . udp 17 false 512" NOERROR - 0 30.00056859s
[ERROR] plugin/errors: 2 . NS: dial tcp 1.0.0.1:853: i/o timeout
[INFO] 127.0.0.1:46069 - 34633 "NS IN . udp 17 false 512" NOERROR - 0 30.001109111s
[ERROR] plugin/errors: 2 . NS: dial tcp 1.0.0.1:853: i/o timeout
[INFO] 127.0.0.1:52199 - 55341 "NS IN . udp 17 false 512" NOERROR - 0 30.000769912s
[ERROR] plugin/errors: 2 . NS: dial tcp 1.0.0.1:853: i/o timeout
[INFO] 127.0.0.1:44575 - 61374 "NS IN . udp 17 false 512" NOERROR - 0 30.000521003s
[ERROR] plugin/errors: 2 . NS: dial tcp 1.1.1.1:853: i/o timeout
[INFO] 127.0.0.1:59716 - 56622 "NS IN . udp 17 false 512" NOERROR - 0 30.001314379s
[ERROR] plugin/errors: 2 . NS: dial tcp 1.0.0.1:853: i/o timeout
[INFO] 127.0.0.1:43202 - 11517 "NS IN . udp 17 false 512" NOERROR - 0 30.000960835s
[ERROR] plugin/errors: 2 . NS: dial tcp 1.1.1.1:853: i/o timeout
~~~
```


### System Health information

_No response_

### Additional information

_No response_

Same issue here, and there are numerous attempts to dial 1.0.0.1:853 and 1.1.1.1:853.
I tried to had edited /usr/share/hassio/dns.json and /usr/share/hassio/dns/coredns.json, not help with issue, and this issue slowed down my local network.
Pls fix this issue, and let users choose.
```
Feb 24 10:29:33 raspberrypi c25c8d59d674[766]: [ERROR] plugin/errors: 2 . NS: dial tcp 1.1.1.1:853: i/o timeout
Feb 24 10:29:35 raspberrypi c25c8d59d674[766]: [INFO] 127.0.0.1:43745 - 14836 "NS IN . udp 17 false 512" NOERROR - 0 30.001163805s
Feb 24 10:29:35 raspberrypi c25c8d59d674[766]: [ERROR] plugin/errors: 2 . NS: dial tcp 1.1.1.1:853: i/o timeout
Feb 24 10:29:36 raspberrypi c25c8d59d674[766]: [INFO] 127.0.0.1:43857 - 18263 "NS IN . udp 17 false 512" NOERROR - 0 30.000670555s
Feb 24 10:29:36 raspberrypi c25c8d59d674[766]: [ERROR] plugin/errors: 2 . NS: dial tcp 1.1.1.1:853: i/o timeout
Feb 24 10:29:38 raspberrypi c25c8d59d674[766]: [INFO] 127.0.0.1:46192 - 23166 "NS IN . udp 17 false 512" NOERROR - 0 30.000614263s
Feb 24 10:29:38 raspberrypi c25c8d59d674[766]: [ERROR] plugin/errors: 2 . NS: dial tcp 1.0.0.1:853: i/o timeout
Feb 24 10:29:39 raspberrypi c25c8d59d674[766]: [INFO] 127.0.0.1:41931 - 46048 "NS IN . udp 17 false 512" NOERROR - 0 30.000839356s
Feb 24 10:29:39 raspberrypi c25c8d59d674[766]: [ERROR] plugin/errors: 2 . NS: dial tcp 1.1.1.1:853: i/o timeout
Feb 24 10:29:41 raspberrypi c25c8d59d674[766]: [INFO] 127.0.0.1:54422 - 21526 "NS IN . udp 17 false 512" NOERROR - 0 30.00059314s
Feb 24 10:29:41 raspberrypi c25c8d59d674[766]: [ERROR] plugin/errors: 2 . NS: dial tcp 1.0.0.1:853: i/o timeout

--- a/Dockerfile
+++ b/Dockerfile
@@ -18,7 +18,7 @@ RUN \
     && cp -rf plugins/* coredns/plugin/ \
     && cd coredns \
     && sed -i "/^template:template/d" plugin.cfg \
-    && sed -i "/^hosts:.*/i template:template" plugin.cfg \
+    && sed -i "/^hosts:.*/a template:template" plugin.cfg \
     && sed -i "/^forward:.*/i fallback:fallback" plugin.cfg \
     && sed -i "/^hosts:.*/a mdns:mdns" plugin.cfg \
     && sed -i "/route53:route53/d" plugin.cfg \

--- a/Dockerfile
+++ b/Dockerfile
@@ -17,6 +17,8 @@ RUN \
     && git clone --depth 1 -b v${COREDNS_VERSION} https://github.com/coredns/coredns \
     && cp -rf plugins/* coredns/plugin/ \
     && cd coredns \
+    && sed -i "/^template:template/d" plugin.cfg \
+    && sed -i "/^hosts:.*/i template:template" plugin.cfg \
     && sed -i "/^forward:.*/i fallback:fallback" plugin.cfg \
     && sed -i "/^hosts:.*/a mdns:mdns" plugin.cfg \
     && sed -i "/route53:route53/d" plugin.cfg \

--- a/plugins/mdns/mdns.go
+++ b/plugins/mdns/mdns.go
@@ -69,6 +69,7 @@ func (m MDNS) AddARecord(msg *dns.Msg, state *request.Request, name string, addr
 }
 
 func (m MDNS) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	// If systemd-resolved isn't available, plugin can't do anything. Pass to the next one.
 	if m.Resolver == nil {
 		return plugin.NextOrFailure(m.Name(), m.Next, ctx, w, r)
 	}
@@ -105,8 +106,10 @@ func (m MDNS) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (i
 		return dns.RcodeSuccess, nil
 	}
 
-	log.Debugf("No records found for '%s', forwarding to next plugin.", state.QName())
-	return plugin.NextOrFailure(m.Name(), m.Next, ctx, w, r)
+	// Plugin only processes A and AAAA type multicast queries (.local or single name)
+	// Do not forward those to external resolvers. If we don't have an answer no one does
+	log.Debugf("No records found for '%s', returning NXDOMAIN.", state.QName())
+	return dns.RcodeNameError, nil
 }
 
 func GetPrimaryInterface(ctx context.Context, resolver *resolve1.Manager) int32 {

--- a/plugins/mdns/mdns.go
+++ b/plugins/mdns/mdns.go
@@ -100,16 +100,16 @@ func (m MDNS) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (i
 		// There may be uncommon errors though so not swallowing it while debugging
 		log.Debug(err)
 
-	} else if m.AddARecord(msg, &state, hostName, addresses) {
+	} else {
+		m.AddARecord(msg, &state, hostName, addresses)
 		log.Debug(msg)
-		w.WriteMsg(msg)
-		return dns.RcodeSuccess, nil
 	}
 
 	// Plugin only processes A and AAAA type multicast queries (.local or single name)
-	// Do not forward those to external resolvers. If we don't have an answer no one does
-	log.Debugf("No records found for '%s', returning NXDOMAIN.", state.QName())
-	return dns.RcodeNameError, nil
+	// Whether an answer was found or not this is end of the line, do not forward to external resolvers
+	// Always return NOERROR since we are not authoritative for this domain
+	w.WriteMsg(msg)
+	return dns.RcodeSuccess, nil
 }
 
 func GetPrimaryInterface(ctx context.Context, resolver *resolve1.Manager) int32 {

--- a/rootfs/usr/share/tempio/corefile
+++ b/rootfs/usr/share/tempio/corefile
@@ -11,6 +11,9 @@
     template ANY AAAA local.hass.io hassio {
         rcode NOERROR
     }
+    template ANY A local.hass.io hassio {
+        rcode NXDOMAIN
+    }
     mdns
     forward . {{ join " " .servers }} {{ if len .locals | eq 0 }}dns://127.0.0.11{{ else }}{{ join " " .locals }}{{ end }} {
         except local.hass.io

